src/module.c

```diff
@@ -5696,37 +5699,64 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
     if ((cmd->arity > 0 && cmd->arity != argc) || (argc < -cmd->arity)) {
         errno = EINVAL;
         if (error_as_call_replies) {
-            sds msg = sdscatfmt(sdsempty(), "ERR Wrong number of "
-                                            "args calling Redis command %S.", c->argv[0]->ptr);
+            sds msg = sdscatfmt(sdsempty(), "Wrong number of "
+                                            "args calling Redis command '%S'.", c->cmd->fullname);
             reply = callReplyCreateError(msg, ctx);
         }
         goto cleanup;
     }
 
-    if (flags & REDISMODULE_ARGV_SAFE_MODE) {
-        /* Basically on safe mode we want to only allow commands that can
+    if (flags & REDISMODULE_ARGV_SCRIPT_MODE) {
+        /* Basically on script mode we want to only allow commands that can
          * be executed on scripts (CMD_NOSCRIPT is not set on the command flags) */
         if (cmd->flags & CMD_NOSCRIPT) {
             errno = EACCES;
             if (error_as_call_replies) {
-                sds msg = sdscatfmt(sdsempty(), "ERR Unsafe command "
-                                                "%S was called.", c->argv[0]->ptr);
+                sds msg = sdscatfmt(sdsempty(), "command '%S' is not allowed on script mode", c->cmd->fullname);
                 reply = callReplyCreateError(msg, ctx);
             }
             goto cleanup;
         }
     }
 
-    if ((flags & REDISMODULE_ARGV_NO_WRITES) && (cmd->flags & CMD_WRITE)) {
-        if (cmd->flags & CMD_WRITE) {
+    if (cmd->flags & CMD_WRITE) {
+        if (flags & REDISMODULE_ARGV_NO_WRITES) {
             errno = EACCES;
             if (error_as_call_replies) {
-                sds msg = sdscatfmt(sdsempty(), "ERR Write command %S was "
-                                                "called while write is not allowed.", c->argv[0]->ptr);
+                sds msg = sdscatfmt(sdsempty(), "Write command '%S' was "
+                                                "called while write is not allowed.", c->cmd->fullname);
                 reply = callReplyCreateError(msg, ctx);
             }
             goto cleanup;
         }
+
+        if (flags & REDISMODULE_ARGV_SCRIPT_MODE) {
+            /* on script mode, if a command is a write command,
+             * We will not run it if we encounter disk error
+             * or we do not have enough replicas */
+
+            if (!checkGoodReplicasStatus())
+            {
+                errno = EACCES;
```

Comment:
the docs says `EACCES` is for ACL, let's use something else for all the new usages, and document it.

Reply:
Fixed, set:
ENOSPC: Write command is not allowed
ESPIPE: Command not allowed on script mode